### ISSUE TITLE
Concurrent map writes in v16.0.0-rc.4

### ISSUE THREAD

**Issue by bfreis** [thought-machine/please#1752]
I'm getting this issue when running please v16.0.0-rc.4. It doesn't seem to be deterministic, suggesting a data race somewhere:

```
fatal error: concurrent map writes
fatal error: concurrent map writes

goroutine 41 [running]:
runtime.throw(0x4abe554, 0x15)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/runtime/panic.go:1117 +0x72 fp=0xc0009bb8b8 sp=0xc0009bb888 pc=0x4038712
runtime.mapassign_faststr(0x4962000, 0xc0001080f0, 0x4ab1b25, 0x7, 0x51d6f60)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/runtime/map_faststr.go:291 +0x3d8 fp=0xc0009bb920 sp=0xc0009bb8b8 pc=0x4015218
text/template.(*Template).associate(0xc000102040, 0xc000102040, 0xc0009c8c60, 0xc0006f2b01)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/text/template/template.go:227 +0xba fp=0xc0009bb960 sp=0xc0009bb920 pc=0x43f0a7a
text/template.(*Template).AddParseTree(0xc000102040, 0x4ab1b25, 0x7, 0xc0009c8c60, 0x0, 0x0, 0x0)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/text/template/template.go:133 +0x105 fp=0xc0009bb9b0 sp=0xc0009bb960 pc=0x43ef6a5
text/template.(*Template).Parse(0xc000102040, 0x4ae99f4, 0x234, 0x203000, 0x203000, 0x0)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/text/template/template.go:208 +0x908 fp=0xc0009bbac0 sp=0xc0009bb9b0 pc=0x43f0608
github.com/thought-machine/please/src/parse.GetInternalPackage(0xc000596000, 0xc0005997a0, 0x0, 0x0, 0xc0001cc468)
        src/parse/internal_package.go:45 +0x57 fp=0xc0009bbbd8 sp=0xc0009bbac0 pc=0x46958d7
github.com/thought-machine/please/src/parse.parsePackage(0xc00033f340, 0x4ab9672, 0x7, 0x4ab967a, 0x6, 0xc0003336a8, 0x6, 0xc0000f852c, 0x13, 0xc0006eebd1, ...)
        src/parse/parse_step.go:188 +0xbe5 fp=0xc0009bbd68 sp=0xc0009bbbd8 pc=0x46986e5
github.com/thought-machine/please/src/parse.parse(0x2, 0xc00033f340, 0x4ab9672, 0x7, 0x4ab967a, 0x6, 0xc0003336a8, 0x6, 0xc0000f852c, 0x13, ...)
        src/parse/parse_step.go:59 +0x212 fp=0xc0009bbe48 sp=0xc0009bbd68 pc=0x4695f92
github.com/thought-machine/please/src/parse.Parse(0x2, 0xc00033f340, 0x4ab9672, 0x7, 0x4ab967a, 0x6, 0xc0003336a8, 0x6, 0xc0000f852c, 0x13, ...)
        src/parse/parse_step.go:33 +0xa7 fp=0xc0009bbee0 sp=0xc0009bbe48 pc=0x4695cc7
github.com/thought-machine/please/src/plz.doTasks.func1(0x2, 0xc00033f340, 0x4ab9672, 0x7, 0x4ab967a, 0x6, 0xc0003336a8, 0x6, 0xc0000f852c, 0x13, ...)
        src/plz/plz.go:85 +0x99 fp=0xc0009bbf68 sp=0xc0009bbee0 pc=0x4806bf9
runtime.goexit()
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/runtime/asm_amd64.s:1371 +0x1 fp=0xc0009bbf70 sp=0xc0009bbf68 pc=0x40729c1
created by github.com/thought-machine/please/src/plz.doTasks
        src/plz/plz.go:84 +0x438

goroutine 1 [semacquire]:
sync.runtime_Semacquire(0xc0006eeac8)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/runtime/sema.go:56 +0x45
sync.(*WaitGroup).Wait(0xc0006eeac0)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/sync/waitgroup.go:130 +0x65
github.com/thought-machine/please/src/plz.Run(0xc0005802d0, 0x1, 0x1, 0x0, 0x0, 0x0, 0xc00033f340, 0xc000596000, 0x4ab093d, 0x6, ...)
        src/plz/plz.go:59 +0x486
main.runPlease(0xc00033f340, 0xc0005802d0, 0x1, 0x1)
        src/please.go:899 +0x25d
main.Please(0xc0005802d0, 0x1, 0x1, 0xc000596000, 0x1, 0xc0001ca400, 0x0)
        src/please.go:872 +0x433
main.runBuild(0xc0005802d0, 0x1, 0x1, 0x1, 0xc000393d98, 0x43c5914)
        src/please.go:955 +0xba
main.glob..func5(0x4962780)
        src/please.go:442 +0x9e
main.execute(0x4966925, 0x3, 0x0)
        src/please.go:1145 +0xe9
main.main()
        src/please.go:1149 +0x5d

goroutine 19 [chan receive]:
github.com/thought-machine/please/src/cli.handleSignals()
        src/cli/process.go:20 +0x145
created by github.com/thought-machine/please/src/cli.init.0
        src/cli/process.go:12 +0x35

goroutine 34 [syscall]:
os/signal.signal_recv(0x0)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/runtime/sigqueue.go:165 +0x9d
os/signal.loop()
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/os/signal/signal_unix.go:23 +0x25
created by os/signal.Notify.func1.1
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/os/signal/signal.go:151 +0x45

goroutine 6 [chan receive]:
github.com/golang/glog.(*loggingT).flushDaemon(0x51a45e0)
        _build/github.com/golang/glog/glog.go:883 +0x8b
created by github.com/golang/glog.init.0
        _build/github.com/golang/glog/glog.go:410 +0x274

goroutine 10 [runnable]:
syscall.syscall(0x4089280, 0xa, 0xc0002992d0, 0x0, 0x0, 0x0, 0x0)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/runtime/sys_darwin.go:19 +0x2e
syscall.Fstat(0xa, 0xc0002992d0, 0x406d728, 0xc000182c00)
        /Users/distiller/project/plz-out/bin/third_party/go/toolc:1969 +0x5c
os.newFile.func1(...)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/os/file_unix.go:147
os.ignoringEINTR(...)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/os/file_posix.go:245
os.newFile(0xa, 0xc0003ca230, 0x44, 0x1, 0xa)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/os/file_unix.go:146 +0x152
os.openFileNolog(0xc0003ca230, 0x44, 0x0, 0xc000000000, 0xc0002994d8, 0x451080e, 0xc0003ca26c)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/os/file_unix.go:240 +0x1d7
os.OpenFile(0xc0003ca230, 0x44, 0x0, 0x0, 0xc0000c00a0, 0xc0003ca26c, 0x8)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/os/file.go:331 +0x65
os.Open(...)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/os/file.go:311
github.com/karrick/godirwalk.readdirents(0xc0003ca230, 0x44, 0xc0001da000, 0x10000, 0x10000, 0x44, 0x4923700, 0x1, 0xc0003ca230, 0xc000299650)
        _build/github.com/karrick/godirwalk/readdir_unix.go:22 +0x5e
github.com/karrick/godirwalk.ReadDirents(...)
        _build/github.com/karrick/godirwalk/readdir.go:93
github.com/karrick/godirwalk.walk(0xc0003ca230, 0x44, 0xc0000a7170, 0xc000299d48, 0x44, 0x0)
        _build/github.com/karrick/godirwalk/walk.go:260 +0x24c
github.com/karrick/godirwalk.walk(0xc0001ee180, 0x3b, 0xc0000a6ca8, 0xc000299d48, 0x3b, 0x0)
        _build/github.com/karrick/godirwalk/walk.go:275 +0x378
github.com/karrick/godirwalk.walk(0xc0001ee100, 0x32, 0xc0000a6c30, 0xc000299d48, 0x32, 0xc0000b02a0)
        _build/github.com/karrick/godirwalk/walk.go:275 +0x378
github.com/karrick/godirwalk.walk(0xc0003b83c0, 0x2e, 0xc0000a6960, 0xc000299d48, 0x2e, 0x0)
        _build/github.com/karrick/godirwalk/walk.go:275 +0x378
github.com/karrick/godirwalk.walk(0xc000041dd0, 0x23, 0xc0000a6030, 0xc000299d48, 0x0, 0x0)
        _build/github.com/karrick/godirwalk/walk.go:275 +0x378
github.com/karrick/godirwalk.Walk(0xc000041dd0, 0x23, 0xc000068d48, 0xc0005a4000, 0x0)
        _build/github.com/karrick/godirwalk/walk.go:195 +0x16e
github.com/thought-machine/please/src/fs.WalkMode(0xc000041dd0, 0x23, 0xc0005b23f0, 0x1, 0xc0000b0040)
        src/fs/walk.go:27 +0xfb
github.com/thought-machine/please/src/fs.Walk(0xc000041dd0, 0x23, 0xc0000b0040, 0x0, 0x0)
        src/fs/walk.go:13 +0x6f
github.com/thought-machine/please/src/cache.(*dirCache).clean(0xc0000c00a0, 0x280000000, 0x200000000, 0x0)
        src/cache/dir_cache.go:395 +0xe5
created by github.com/thought-machine/please/src/cache.newDirCache
        src/cache/dir_cache.go:361 +0x20a

goroutine 11 [chan receive]:
github.com/thought-machine/please/src/cache.(*asyncCache).run(0xc0005808a0)
        src/cache/async_cache.go:68 +0x145
created by github.com/thought-machine/please/src/cache.newAsyncCache
        src/cache/async_cache.go:34 +0xdd

goroutine 12 [chan receive]:
github.com/thought-machine/please/src/cache.(*asyncCache).run(0xc0005808a0)
        src/cache/async_cache.go:68 +0x145
created by github.com/thought-machine/please/src/cache.newAsyncCache
        src/cache/async_cache.go:34 +0xdd

goroutine 13 [chan receive]:
github.com/thought-machine/please/src/cache.(*asyncCache).run(0xc0005808a0)
        src/cache/async_cache.go:68 +0x145
created by github.com/thought-machine/please/src/cache.newAsyncCache
        src/cache/async_cache.go:34 +0xdd

goroutine 14 [chan receive]:
github.com/thought-machine/please/src/cache.(*asyncCache).run(0xc0005808a0)
        src/cache/async_cache.go:68 +0x145
created by github.com/thought-machine/please/src/cache.newAsyncCache
        src/cache/async_cache.go:34 +0xdd

goroutine 15 [chan receive]:
github.com/thought-machine/please/src/cache.(*asyncCache).run(0xc0005808a0)
        src/cache/async_cache.go:68 +0x145
created by github.com/thought-machine/please/src/cache.newAsyncCache
        src/cache/async_cache.go:34 +0xdd

goroutine 16 [chan receive]:
github.com/thought-machine/please/src/cache.(*asyncCache).run(0xc0005808a0)
        src/cache/async_cache.go:68 +0x145
created by github.com/thought-machine/please/src/cache.newAsyncCache
        src/cache/async_cache.go:34 +0xdd

goroutine 50 [runnable]:
sync.runtime_SemacquireMutex(0xc0001cc474, 0xc000580300, 0x0)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/runtime/sema.go:71 +0x47
sync.(*RWMutex).RLock(...)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/sync/rwmutex.go:63
github.com/thought-machine/please/src/core.(*BuildGraph).Target(0xc0001cc440, 0x4ab9672, 0x7, 0x4ab967a, 0x6, 0x0, 0x0, 0x0)
        src/core/graph.go:73 +0xfe
github.com/thought-machine/please/src/output.processResult(0xc00033f340, 0xc000801a40, 0xc000180d80, 0x6, 0x6, 0xc00033f300, 0xc000294c60, 0xc000294c78, 0xc000294cc0, 0xc00000e018, ...)
        src/output/shell_output.go:179 +0x108
github.com/thought-machine/please/src/output.MonitorState(0x4c2ffb8, 0xc00030a000, 0xc00033f340, 0x0, 0x0, 0x0)
        src/output/shell_output.go:81 +0x32c
main.runPlease.func1(0x4c2ffb8, 0xc0001cc480, 0xc00033f340, 0xc000000001, 0xc0001be930)
        src/please.go:896 +0x79
created by main.runPlease
        src/please.go:895 +0x1c5

goroutine 66 [select]:
github.com/thought-machine/please/src/out(0xc000295ee0, 0x4c2ffb8, 0xc00030a000)
        src/output/interactive_display.go:55 +0x286
github.com/thought-machine/please/src/output.display(0x4c2ffb8, 0xc00030a000, 0xc00033f340, 0xc000180d80, 0x6, 0x6)
        src/output/interactive_display.go:43 +0x178
github.com/thought-machine/please/src/output.MonitorState.func1(0x4c2ff00, 0x4c2ffb8, 0xc00030a000, 0xc00033f340, 0xc000180d80, 0x6, 0x6, 0xc0005a6000)
        src/output/shell_output.go:70 +0xcf
created by github.com/thought-machine/please/src/output.MonitorState
        src/output/shell_output.go:66 +0x22c

goroutine 51 [chan receive]:
github.com/thought-machine/please/src/core.(*BuildState).UpdateResources(0xc00033f340)
        src/core/resources.go:36 +0x17f
created by github.com/thought-machine/please/src/plz.Run
        src/plz/plz.go:34 +0x685

goroutine 24 [running]:
        goroutine running on other thread; stack unavailable
created by github.com/thought-machine/please/src/plz.doTasks
        src/plz/plz.go:84 +0x438

goroutine 53 [semacquire]:
sync.runtime_SemacquireMutex(0xc0000c003c, 0x4006900, 0x1)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/runtime/sema.go:71 +0x47
sync.(*Mutex).lockSlow(0xc0000c0038)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/sync/mutex.go:138 +0x105
sync.(*Mutex).Lock(...)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/sync/mutex.go:81
github.com/Workiva/go-datastructures/queue.(*PriorityQueue).Get(0xc0000c0000, 0x1, 0xc0002a7a70, 0x1, 0x1, 0x0, 0x0)
        _build/github.com/Workiva/go-datastructures/queue/priority_queue.go:164 +0x4b5
github.com/thought-machine/please/src/core.(*BuildState).feedQueues(0xc00033f340, 0xc0001d4240, 0xc0001d42a0, 0xc0001d4300, 0xc0001d4360, 0xc0001d43c0)
        src/core/state.go:320 +0x73
created by github.com/thought-machine/please/src/core.(*BuildState).TaskQueues
        src/core/state.go:310 +0x125

goroutine 54 [runnable]:
syscall.syscall(0x4089100, 0x9, 0xc000116aa0, 0x48, 0x48, 0x0, 0x0)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/runtime/sys_darwin.go:19 +0x2e
syscall.write(0x9, 0xc000116aa0, 0x48, 0xa0, 0x0, 0x1, 0xc00029b198)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/syscall/zsyscall_darwin_amd64.go:1734 +0x66
syscall.Write(...)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/syscall/syscall_unix.go:212
internal/poll.ignoringEINTRIO(...)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/internal/poll/fd_unix.go:581
internal/poll.(*FD).Write(0xc000107980, 0xc000116aa0, 0x48, 0xa0, 0x0, 0x0, 0x0)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/internal/poll/fd_unix.go:274 +0x13a
os.(*File).write(...)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/os/file_posix.go:48
os.(*File).Write(0xc000010e20, 0xc000116aa0, 0x48, 0xa0, 0x51a4560, 0x0, 0x0)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/os/file.go:174 +0x8e
log.(*Logger).Output(0xc000104910, 0x7, 0xc00090c050, 0x47, 0x0, 0x0)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/log/log.go:184 +0x284
gopkg.in/op/go-logging%2ev1.(*LogBackend).Log(0xc0005808d0, 0x5, 0x5, 0xc0005afb00, 0x90, 0x4385cc0)
        _build/gopkg.in/op/go-logging.v1/log_nix.go:75 +0x2ba
gopkg.in/op/go-logging%2ev1.(*backendFormatter).Log(0xc00043dc80, 0x5, 0x4, 0xc0005afa70, 0xc000453c28, 0xc0005afa01)
        _build/gopkg.in/op/go-logging.v1/format.go:399 +0xbe
gopkg.in/op/go-logging%2ev1.(*moduleLeveled).Log(0xc000103e40, 0x5, 0x3, 0xc0005afa70, 0xc0005af901, 0x1ba183b38b8e0)
        _build/gopkg.in/op/go-logging.v1/level.go:116 +0xeb
gopkg.in/op/go-logging%2ev1.(*multiLogger).Log(0xc00000f2d8, 0x5, 0x2, 0xc0005af9e0, 0xc00029b501, 0x400ee78)
        _build/gopkg.in/op/go-logging.v1/multi.go:31 +0x18a
gopkg.in/op/go-logging%2ev1.(*Logger).log(0xc000238b40, 0x5, 0x4ace346, 0x24, 0xc000328210, 0x1, 0x1)
        _build/gopkg.in/op/go-logging.v1/logger.go:162 +0x1cd
gopkg.in/op/go-logging%2ev1.(*Logger).Debug(...)
        _build/gopkg.in/op/go-logging.v1/logger.go:239
github.com/thought-machine/please/src/build.buildTarget(0x0, 0xc00033f340, 0xc0009c6a80, 0xc000332200, 0x0, 0x0)
        src/build/build_step.go:205 +0x26e5
github.com/thought-machine/please/src/build.Build(0x0, 0xc00033f340, 0xc000038bd0, 0x0, 0xc0003322c1, 0xc, 0x0, 0x0, 0xc0000f8500)
        src/build/build_step.go:47 +0xcf
github.com/thought-machine/please/src/plz.doTasks(0x0, 0xc00033f340, 0xc0001d4240, 0xc0001d42a0, 0xc0001d4300, 0x0)
        src/plz/plz.go:93 +0x32c
github.com/thought-machine/please/src/plz.Run.func1(0xc00033f340, 0xc0001d4240, 0xc0001d42a0, 0xc0001d4300, 0xc0006eeac0, 0x0)
        src/plz/plz.go:48 +0x58
created by github.com/thought-machine/please/src/plz.Run
        src/plz/plz.go:47 +0x397

goroutine 55 [select]:
github.com/thought-machine/please/src/plz.doTasks(0x1, 0xc00033f340, 0xc0001d4240, 0xc0001d42a0, 0xc0001d4300, 0x0)
        src/plz/plz.go:78 +0x185
github.com/thought-machine/please/src/plz.Run.func1(0xc00033f340, 0xc0001d4240, 0xc0001d42a0, 0xc0001d4300, 0xc0006eeac0, 0x1)
        src/plz/plz.go:48 +0x58
created by github.com/thought-machine/please/src/plz.Run
        src/plz/plz.go:47 +0x397

goroutine 56 [select]:
github.com/thought-machine/please/src/plz.doTasks(0x2, 0xc00033f340, 0xc0001d4240, 0xc0001d42a0, 0xc0001d4300, 0x0)
        src/plz/plz.go:78 +0x185
github.com/thought-machine/please/src/plz.Run.func1(0xc00033f340, 0xc0001d4240, 0xc0001d42a0, 0xc0001d4300, 0xc0006eeac0, 0x2)
        src/plz/plz.go:48 +0x58
created by github.com/thought-machine/please/src/plz.Run
        src/plz/plz.go:47 +0x397

goroutine 57 [select]:
github.com/thought-machine/please/src/plz.doTasks(0x3, 0xc00033f340, 0xc0001d4240, 0xc0001d42a0, 0xc0001d4300, 0x0)
        src/plz/plz.go:78 +0x185
github.com/thought-machine/please/src/plz.Run.func1(0xc00033f340, 0xc0001d4240, 0xc0001d42a0, 0xc0001d4300, 0xc0006eeac0, 0x3)
        src/plz/plz.go:48 +0x58
created by github.com/thought-machine/please/src/plz.Run
        src/plz/plz.go:47 +0x397

goroutine 58 [select]:
github.com/thought-machine/please/src/plz.doTasks(0x4, 0xc00033f340, 0xc0001d4240, 0xc0001d42a0, 0xc0001d4300, 0x4072900)
        src/plz/plz.go:78 +0x185
github.com/thought-machine/please/src/plz.Run.func1(0xc00033f340, 0xc0001d4240, 0xc0001d42a0, 0xc0001d4300, 0xc0006eeac0, 0x4)
        src/plz/plz.go:48 +0x58
created by github.com/thought-machine/please/src/plz.Run
        src/plz/plz.go:47 +0x397

goroutine 59 [select]:
github.com/thought-machine/please/src/plz.doTasks(0x5, 0xc00033f340, 0xc0001d4240, 0xc0001d42a0, 0xc0001d4300, 0x0)
        src/plz/plz.go:78 +0x185
github.com/thought-machine/please/src/plz.Run.func1(0xc00033f340, 0xc0001d4240, 0xc0001d42a0, 0xc0001d4300, 0xc0006eeac0, 0x5)
        src/plz/plz.go:48 +0x58
created by github.com/thought-machine/please/src/plz.Run
        src/plz/plz.go:47 +0x397

goroutine 60 [runnable]:
sync.runtime_Semacquire(0xc000333c58)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/runtime/sema.go:56 +0x45
sync.(*WaitGroup).Wait(0xc000333c50)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/sync/waitgroup.go:130 +0x65
github.com/Workiva/go-datastructures/queue.(*PriorityQueue).Put(0xc0000c0000, 0xc0009b75c0, 0x1, 0x1, 0x0, 0x0)
        _build/github.com/Workiva/go-datastructures/queue/priority_queue.go:147 +0x185
github.com/thought-machine/please/src/core.(*BuildState).AddPendingParse(0xc00033f340, 0x4ab9672, 0x7, 0x4ab967a, 0x6, 0xc0003336a8, 0x6, 0xc0000f852c, 0x13, 0xc0006eebd1, ...)
        src/core/state.go:278 +0x257
github.com/thought-machine/please/src/core.(*BuildState).queueTarget(0xc00033f340, 0x4ab9672, 0x7, 0x4ab967a, 0x6, 0xc0003336a8, 0x6, 0xc0000f852c, 0x13, 0xc0006eebd1, ...)
        src/core/state.go:858 +0x8cc
github.com/thought-machine/please/src/core.(*BuildState).queueTarget(0xc00033f340, 0xc0000f852c, 0x13, 0xc0006eebd1, 0x7, 0xc0003336a8, 0x6, 0xc000038bd0, 0x0, 0xc000332ed2, ...)
        src/core/state.go:910 +0x330
github.com/thought-machine/please/src/core.(*BuildState).QueueTarget(...)
        src/core/state.go:850
github.com/thought-machine/please/src/parse.activateTarget(0x2, 0xc00033f340, 0xc0002a0a20, 0xc0000f852c, 0x13, 0xc0000f8538, 0x7, 0xc0000f8524, 0x6, 0xc000038bd0, ...)
        src/parse/parse_step.go:170 +0x6fb
github.com/thought-machine/please/src/parse.parse(0x2, 0xc00033f340, 0xc0000f852c, 0x13, 0xc0000f8538, 0x7, 0xc0000f8524, 0x6, 0xc000038bd0, 0x0, ...)
        src/parse/parse_step.go:65 +0x33c
github.com/thought-machine/please/src/parse.Parse(0x2, 0xc00033f340, 0xc0000f852c, 0x13, 0xc0000f8538, 0x7, 0xc0000f8524, 0x6, 0xc000038bd0, 0x0, ...)
        src/parse/parse_step.go:33 +0xa7
github.com/thought-machine/please/src/plz.doTasks.func1(0x2, 0xc00033f340, 0xc0000f852c, 0x13, 0xc0000f8538, 0x7, 0xc0000f8524, 0x6, 0xc000038bd0, 0x0, ...)
        src/plz/plz.go:85 +0x99
created by github.com/thought-machine/please/src/plz.doTasks
        src/plz/plz.go:84 +0x438

goroutine 62 [runnable]:
text/template/parse.(*lexer).emit(...)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/text/template/parse/lex.go:157
text/template/parse.lexText(0xc0006df600, 0x4b48c20)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/text/template/parse/lex.go:269 +0x4fd
text/template/parse.(*lexer).run(0xc0006df600)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/text/template/parse/lex.go:230 +0x37
created by text/template/parse.lex
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/text/template/parse/lex.go:223 +0x14b

goroutine 38 [chan receive]:
github.com/thought-machine/please/src/core.(*BuildState).WaitForBuiltTarget(0xc00033f340, 0xc000038bd0, 0x0, 0xc000332c51, 0xb, 0x0, 0x0, 0xc0000f85af, 0x0, 0x4aae355, ...)
        src/core/state.go:771 +0x46c
github.com/thought-machine/please/src/core.(*BuildState).WaitForBuiltTarget(0xc00033f340, 0xc000038bd0, 0x0, 0xc000332c51, 0xb, 0x0, 0x0, 0xc0000f85af, 0x0, 0x4aae355, ...)
        src/core/state.go:786 +0x345
github.com/thought-machine/please/src/parse.parse(0x0, 0xc00033f340, 0xc0000f85af, 0x0, 0xc0000f85b2, 0x6, 0xc0000f85a4, 0xb, 0xc000038bd0, 0x0, ...)
        src/parse/parse_step.go:53 +0x411
github.com/thought-machine/please/src/parse.Parse(0x0, 0xc00033f340, 0xc0000f85af, 0x0, 0xc0000f85b2, 0x6, 0xc, 0xb, 0xc000038bd0, 0x0, ...)
        src/parse/parse_step.go:33 +0xa7
github.com/thought-machine/please/src/plz.doTasks.func1(0x0, 0xc00033f340, 0xc0000f85af, 0x0, 0xc0000f85b2, 0x6, 0xc0000f85a4, 0xb, 0xc000038bd0, 0x0, ...)
        src/plz/plz.go:85 +0x99
created by github.com/thought-machine/please/src/plz.doTasks
        src/plz/plz.go:84 +0x438

goroutine 42 [runnable]:
text/template/parse.(*lexer).emit(...)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/text/template/parse/lex.go:157
text/template/parse.lexText(0xc0006f0500, 0x4b48c20)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/text/template/parse/lex.go:269 +0x4fd
text/template/parse.(*lexer).run(0xc0006f0500)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/text/template/parse/lex.go:230 +0x37
created by text/template/parse.lex
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/text/template/parse/lex.go:223 +0x14b

goroutine 24 [running]:
runtime.throw(0x4abe554, 0x15)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/runtime/panic.go:1117 +0x72 fp=0xc0008c18b8 sp=0xc0008c1888 pc=0x4038712
runtime.mapassign_faststr(0x4962000, 0xc0001080f0, 0x4ab1b25, 0x7, 0x51d6f60)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/runtime/map_faststr.go:291 +0x3d8 fp=0xc0008c1920 sp=0xc0008c18b8 pc=0x4015218
text/template.(*Template).associate(0xc000102040, 0xc000102040, 0xc0008459e0, 0xc000841d01)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/text/template/template.go:227 +0xba fp=0xc0008c1960 sp=0xc0008c1920 pc=0x43f0a7a
text/template.(*Template).AddParseTree(0xc000102040, 0x4ab1b25, 0x7, 0xc0008459e0, 0x0, 0x0, 0x0)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/text/template/template.go:133 +0x105 fp=0xc0008c19b0 sp=0xc0008c1960 pc=0x43ef6a5
text/template.(*Template).Parse(0xc000102040, 0x4ae99f4, 0x234, 0x203000, 0x203000, 0x57b6808)
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/text/template/template.go:208 +0x908 fp=0xc0008c1ac0 sp=0xc0008c19b0 pc=0x43f0608
github.com/thought-machine/please/src/parse.GetInternalPackage(0xc000596000, 0xc0002a0bd0, 0xc000290ce0, 0xc000183680, 0xc000290c80)
        src/parse/internal_package.go:45 +0x57 fp=0xc0008c1bd8 sp=0xc0008c1ac0 pc=0x46958d7
github.com/thought-machine/please/src/parse.parsePackage(0xc00033f340, 0x4ab9672, 0x7, 0x4ab967a, 0x6, 0x0, 0x0, 0xc000038bd0, 0x0, 0xc000332c51, ...)
        src/parse/parse_step.go:188 +0xbe5 fp=0xc0008c1d68 sp=0xc0008c1bd8 pc=0x46986e5
github.com/thought-machine/please/src/parse.parse(0x5, 0xc00033f340, 0x4ab9672, 0x7, 0x4ab967a, 0x6, 0x0, 0x0, 0xc000038bd0, 0x0, ...)
        src/parse/parse_step.go:59 +0x212 fp=0xc0008c1e48 sp=0xc0008c1d68 pc=0x4695f92
github.com/thought-machine/please/src/parse.Parse(0x5, 0xc00033f340, 0x4ab9672, 0x7, 0x4ab967a, 0x6, 0x0, 0x0, 0xc000038bd0, 0x0, ...)
        src/parse/parse_step.go:33 +0xa7 fp=0xc0008c1ee0 sp=0xc0008c1e48 pc=0x4695cc7
github.com/thought-machine/please/src/plz.doTasks.func1(0x5, 0xc00033f340, 0x4ab9672, 0x7, 0x4ab967a, 0x6, 0x0, 0x0, 0xc000038bd0, 0x0, ...)
        src/plz/plz.go:85 +0x99 fp=0xc0008c1f68 sp=0xc0008c1ee0 pc=0x4806bf9
runtime.goexit()
        /Users/distiller/project/plz-out/bin/third_party/go/toolchain/src/runtime/asm_amd64.s:1371 +0x1 fp=0xc0008c1f70 sp=0xc0008c1f68 pc=0x40729c1
created by github.com/thought-machine/please/src/plz.doTasks
        src/plz/plz.go:84 +0x438
```

**Comment by peterebden** [comment]
Hmmmmm this is an odd one - I can see what it's doing, and there are a couple of ways we could protect it, but it should only be entering this function once anyway. We've certainly not seen this internally.

Can you shed any light on what your repo looks like? Do you have any subrepos configured, by any chance? (That's the most obvious way I can see for this to be happening)

**Comment by bfreis** [comment]
Hi,

Yes, I was playing with subrepos. I was trying to define a local subrepo
pointing to another directory. I ran into all sorts of problems, some that
I found mentioned in Gitter back in 2019 but couldn't find any solutions. I
think what I ended up doing was creating a symlink in plz-out/gen to the
other repo - which turned out to be a bad idea as it polluted the other
repo with what seems like caches. When I did that, the local subrepo
worked, but I got this concurrency issue randomly.
